### PR TITLE
Add spectral clustering

### PR DESCRIPTION
**Pull request by RicardoSantos-99**
Add spectral clustering

Implements Scholar.Cluster.SpectralClustering via the normalized graph
Laplacian. The affinity matrix is built with an RBF kernel (gamma
configurable) or accepted as precomputed. Self-loops are discarded before
computing degrees, matching scipy's csgraph convention. Isolated vertices
keep a unit diagonal, matching scikit-learn's _set_diag behavior.

The spectral embedding takes the k eigenvectors of the smallest eigenvalues
of the Laplacian (eigh output is not sorted, so an explicit argsort is
needed), rescales them as u = D^-1/2 v per the relaxed Ncut formulation,
and applies a deterministic sign flip. K-means then runs on this embedding.

The struct exposes both labels and the embedding itself.

Related to #135.

### CODE REVIEW

**Comment by krstopro** [comment]
I remember giving this one a try long time ago, but one of the issues was that there were no efficient solvers for top-k eigenvalues in Nx. Is it still the case?

**Comment by RicardoSantos-99** [comment]
> I remember giving this one a try long time ago, but one of the issues was that there were no efficient solvers for top-k eigenvalues in Nx. Is it still the case?

Yeah, I think that's still the case. As far as I can tell Nx only has the
dense eigh, I couldn't find any partial or iterative solver in there.

So right now this just runs the full eigh and takes the k smallest
eigenvectors, which ends up O(N^3) / O(N^2) (I tried to be upfront about
that in the moduledoc). My thinking was that it matches sklearn's dense
eigen_solver and the values line up with it, and it felt in line with the
other exact/dense stuff already in Scholar like exact t-SNE. But I might be
missing something.

Would you be ok merging the dense version for now and revisiting if a
partial eigensolver ever lands in Nx? Or do you think it's better to hold
off until there's a proper top-k solver? Happy to go either way.

**Comment by krstopro** [comment]
> > I remember giving this one a try long time ago, but one of the issues was that there were no efficient solvers for top-k eigenvalues in Nx. Is it still the case?
> 
> Yeah, I think that's still the case. As far as I can tell Nx only has the dense eigh, I couldn't find any partial or iterative solver in there.
> 
> So right now this just runs the full eigh and takes the k smallest eigenvectors, which ends up O(N^3) / O(N^2) (I tried to be upfront about that in the moduledoc). My thinking was that it matches sklearn's dense eigen_solver and the values line up with it, and it felt in line with the other exact/dense stuff already in Scholar like exact t-SNE. But I might be missing something.
> 
> Would you be ok merging the dense version for now and revisiting if a partial eigensolver ever lands in Nx? Or do you think it's better to hold off until there's a proper top-k solver? Happy to go either way.

Yes, this seems fine for the first version, especially since you documented the running time.

File: lib/scholar/cluster/spectral_clustering.ex [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        eigh_eps: eigh_eps
+      )
+
+    kmeans =
+      Scholar.Cluster.KMeans.fit(embedding,
+        num_clusters: opts[:num_clusters],
+        num_runs: opts[:num_runs],
+        max_iterations: opts[:max_iterations],
+        key: key
+      )
```
Comment (josevalim) [inline lib/scholar/cluster/spectral_clustering.ex:175]: Ideally we call these two from within `defnp`, so they are both inlined into the same GPU computation. We just returns `{labels, embedding}` from said that `defnp` which we return below.

**Review by josevalim — APPROVED** [review]
One comment and we can ship it (as long as @krstopro agrees!)

**Comment by krstopro** [comment]
> One comment and we can ship it (as long as @krstopro agrees!)

Of course! I don't know why am I not able to approve the pull requests though.

**Comment by josevalim** [comment]
@krstopro weird, you should have permissions to do so!

**Comment by krstopro** [comment]
> @krstopro weird, you should have permissions to do so!

Maybe I should have assigned myself as a reviewer or so. Will figure it out. :)

File: lib/scholar/cluster/spectral_clustering.ex (hunk 1 above)
Comment (RicardoSantos-99) [reply lib/scholar/cluster/spectral_clustering.ex:175]: Done! Moved both calls into a single defnp that returns {labels, embedding}, so everything compiles into one computation now.

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
